drivers/hmis/lib/hmis_data_cleanup/util.rb

```diff
@@ -179,6 +187,93 @@ def self.write_project_unit_summary!(filename: 'hmis_project_summary.csv')
       end
     end
 
+    def self.write_potential_duplicates(filename: 'hmis_potential_duplicates.csv', variant: 'all', full_name: true)
+      Rails.logger.info("Finding potential duplicates (variant: #{variant})")
+
+      data_source_id = GrdaWarehouse::DataSource.hmis.first.id
+
+      # Find all Destination clients that have >1 source client in HMIS
+      destination_id_to_source_ids = GrdaWarehouse::WarehouseClient.where(data_source_id: data_source_id).
+        joins(:source). # drop non existent source clients
+        group(:destination_id).
+        having('count(*) > 1').select('"destination_id", array_agg("source_id") as source_ids').
+        map { |r| [r.destination_id, r.source_ids] }.
+        to_h
+
+      # Map source ID => demographic details
+      source_id_to_info = Hmis::Hud::Client.where(id: destination_id_to_source_ids.values.flatten).
+        where(data_source_id: data_source_id).
+        map do |client|
+          name_parts = if full_name
+            [client.first_name, client.middle_name, client.last_name, client.name_suffix]
+          else
+            [client.first_name, client.last_name]
+          end
+
+          comparison_attrs = {
+            name: name_parts.compact_blank.map(&:strip).join(' ').downcase,
+            dob: client.dob&.strftime('%Y-%m-%d'),
+            ssn: client.ssn,
+            genders: client.gender_multi.excluding(8, 9, 99).sort.map { |k| ::HudUtility2024.gender(k) }.join(', ').presence,
+          }
+          [client.id, comparison_attrs]
+        end.to_h
+
+      rows = []
+      destination_id_to_source_ids.each do |dest_id, source_ids|
+        row = {
+          WarehouseID: dest_id,
+        }
+
+        8.times do |idx|
```

Comment:
Is this just because 8 is a big number?

Reply:
That was the largest number of duplicates in the environment where I was running this